### PR TITLE
Add local caching and new network layer

### PR DESCRIPTION
**Pull request by pdamonkey**
Added new network layer which handles local caching.  
ETag support added (currently used for Sponsors endpoint).  
If a matched Etag is found or a network error is found, the local cache will be used.

### CODE REVIEW

File: SwiftLeeds/Network/URLSession.swift [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            do {
+                let (data, response) = try await self.data(for: request.urlRequest)
+
+                try Task.checkCancellation()
+
+                guard let response = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
+
+                if let eTagKey = request.eTagKey, let eTagValue = response.value(forHTTPHeaderField: "Etag") {
+                    UserDefaults.standard.set(eTagValue, forKey: eTagKey)
+                }
```
Comment (Sherlouk) [inline SwiftLeeds/Network/URLSession.swift:54]: I think we should move the eTag capture to after the `data.write` call.

If we fail to save it, then we'd want to request the data back from origin without eTag, right?

File: SwiftLeeds/Network/URLSession.swift [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+//
+
+import Foundation
+
+public extension URLSession {
+    static var awaitConnectivity: URLSession = {
+        let configuration = URLSessionConfiguration.default
+        configuration.waitsForConnectivity = true
+        configuration.timeoutIntervalForRequest = 5 //60
+        configuration.timeoutIntervalForResource = 5 //60 * 30
```
Comment (Sherlouk) [inline SwiftLeeds/Network/URLSession.swift:15]: are the comments here placeholders of production values?

**Review by Sherlouk — APPROVED** [review]

File: SwiftLeeds/Network/URLSession.swift (hunk 2 above)
Comment (pdamonkey) [reply SwiftLeeds/Network/URLSession.swift:15]: Original timeout values, I'll get rid

File: SwiftLeeds/Network/URLSession.swift (hunk 1 above)
Comment (pdamonkey) [reply SwiftLeeds/Network/URLSession.swift:54]: Good suggestion

Comment (Sherlouk) [reply SwiftLeeds/Network/URLSession.swift:54]: If we capture eTag before we populate cache then there's a journey there which means cache fails, etag populates, on refresh we use eTag, it tells us no new content and to use cache, but cache failed so nothing is ever shown.

eTag must only be used when cache is there as a backup.

**Pull request merged**